Sources/NIOCore/IPAddresses.swift

```diff
@@ -0,0 +1,447 @@
+import CoreFoundation
+
+
+/// `Error` that may be thrown if we fail to create a `IPAddress`
+public enum IPAddressError: Error {
+    /// Given string input is not supported IP Address
+    case failedToParseIPString(String)
+    /// Given string input is not supported IP Address
+    case bytesArrayHasWrongLength(Int)
+}
+
+public typealias IPv4BytesTuple = (UInt8, UInt8, UInt8, UInt8)
+public typealias IPv6BytesTuple = (UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8)
+
+/// Represent the bytes for an `IPv4Address`.
+public struct IPv4Bytes {
+    private let ipv4BytesTuple: IPv4BytesTuple
+    
+    public var bytes: IPv4BytesTuple {
+        return self.ipv4BytesTuple
+    }
+    
+    init(_ bytes: IPv4BytesTuple) {
+        self.ipv4BytesTuple = bytes
+    }
+}
+
+/// Represent the bytes for an `IPv6Address.
+public struct IPv6Bytes {
+    private let ipv6BytesTuple: IPv6BytesTuple
+    
+    public var bytes: IPv6BytesTuple {
+        return self.ipv6BytesTuple
+    }
+    
+    init(_ bytes: IPv6BytesTuple) {
+        self.ipv6BytesTuple = bytes
+    }
+}
+
+/// Represent a single `IPAddress`.
+public enum IPAddress {
+    /// A single IPv4 address for `IPAddress`.
+    public struct IPv4Address {
+        /// The bytes storing the address of the IPv4 address.
+        let address: IPv4Bytes
+        
+        /// Get the `IPv4Address` as a string.
+        var ipAddress: String {
+            self.address.lazy.map({String($0)}).joined(separator: ".")
+        }
+        
+        /// Get the libc address for an IPv4 address.
+        var posix: in_addr {
+            get {
+                return in_addr.init(s_addr: UInt32(self.address.bytes.0) << 24
+                                    + UInt32(self.address.bytes.1) << 16
+                                    + UInt32(self.address.bytes.2) << 8
+                                    + UInt32(self.address.bytes.3))
+            }
+        }
+        
+        /// Creates a new `IPv4Address`.
+        ///
+        /// - parameters:
+        ///   - address: Bytes that hold the IPv4 address.
+        fileprivate init(address: IPv4Bytes) {
+            self.address = address
+        }
+        
+        /// Creates a new `IPv4Address`.
+        ///
+        /// - parameters:
+        ///   - packedBytes: Collection of UInt8 that holds the address.
+        fileprivate init<Bytes: Collection>(packedBytes bytes: Bytes) where Bytes.Element == UInt8, Bytes.Index == Int {
+            self = .init(address: .init((
+                bytes[0], bytes[1], bytes[2], bytes[3]
+            )))
+        }
+        
+        /// Creates a new `IPv4Address`.
+        ///
+        /// - parameters:
+        ///   - string: String representation of an IPv4 address.
+        fileprivate init(string: String) throws {
+            var bytes: [UInt8] = [0,0,0,0]
+            var idx: Int = 0
+            
+            for char in string {
```

Comment:
Let's not compute on the `String` as a collection: it's slow, and it only gives us features we don't want (supporting non-ASCII characters). We can use the `UTF8View` (`string.utf8`) and process the characters as bytes directly.